src/model.ts

```diff
@@ -1,13 +1,30 @@
 import { Request } from 'express';
-import { Readable } from 'stream';
+import { PagingStream } from './paging-stream';
+import { ISearchApiOptions } from './types';
+import { getSearchQueryParams } from './get-search-query-params';
 
 export class HubApiModel {
-  async getStream (request: Request) {
-    console.log(request);
+  private _options: any;
 
-    const ret = new Readable({ read () {} });
-    ret.push({ foo: 'bar' });
-    ret.push(null);
-    return ret;
+  constructor (options: any) {
+    this._options = options;
+  }
+
+  getStream (request: Request) {
+    // TODO revisit how we get these search options
+    const searchOptions: ISearchApiOptions = request.app.locals.searchOptions;
+    const siteUrl = request.app.locals.hubSiteUrl || this._options.defaultSiteUrl;
```

Comment:
I'm not sure how we want to get these configuration options to this provider.

It seemed like the established interface is to accept the Express request object from the output plugin. 

However, it's worth noting that in the case of the catalog export plugins, these configurations won't be coming directly from the request URL, but will rather be assembled by the respective catalog output plugins depending on _their_ requirements.

This led me to this pattern of expecting these options to be attached the request object by the output plugin. But even there, I'm not sure if `app.locals.*` is the right place or just something I picked up from `hub-indexer`.

Input welcome.

Reply:
Good question.  The pattern for Koop to date has been to use the [NPM config package](https://www.npmjs.com/package/config) to manage configuration settings and require it into any file that needs configuration.  I'm not fond of this, and would prefer an `app.locals` solution.  However, this would require that any user of this provider to set `app.locals` in their main server.js file.  This is not necessarily a bad thing, but would need to be documented.  And currently, I don't think this is done in Koop apps built by the CLI (@haoliangyu, can you confirm?), so would require anyone to dig through pre-existing CLI app files and make the adjustment.

Perhaps we can do both and move towards deprecating the use of the `config` package.  You could look for `app.locals.config` and if not found, fallback to sourcing from `config`.

@haoliangyu, any thoughts?

Reply:
Oh my, very sorry, I mis-interpreted your question/comment.  I'm going to delete my previous comment.

Reply:
Let me try this again.  Can you confirm that none of these parameters ultimately come from the request URL?  Even the `q` parameter, which is the search text?

Reply:
From a wholistic perspective, here's what needs to happen for the DCAT feeds
1. domain record is fetched
1. site item is fetched based on the domain record
1. org may be fetched (not necessary right now, but may become so)
1. request is made to Hub Search API
1. results are formatted into a DCAT feed

I was thinking that the DCAT output plugins would perform steps 1-3 and pass a nice config to this search provider. If that's the route we go, I don't think anything will be coming directly from the URL except the hostname used in the domain record lookup.

Talking about this provider in general usage, however, I could definitely see some of them (such as `q`) coming from the URL.

Reply:
Do steps 1-3 happen in the same manner for all output plugins?

Reply:
For now, yes.  The main thing I've heard about that could change the flow somewhat is supporting catalog exports that aren't scoped to a site catalog (for example, export the results of a particular search).

But what I think you're exploring is performing all the steps except that last in the _provider_ as I illustrate in my [original plan](https://confluencewikidev.esri.com/display/Hub/%5BSPIKE%5D+Convert+DCAT+into+a+service?preview=/118686484/118686485/dcat%20service%20sequence%20diagram.png) with a catalog-exports-specific provider.

There are upsides to that idea and we could easily pivot this code to accomplish it, but it seemed like it would also be nice (as Turner has mentioned) to have a generic `hub-search-api` provider that does that one thing: talks to the search API.

So, I'm of two minds but leaning toward pivoting this to a catalog-export-specific plugin since it effectively deduplicates the output providers and makes it extra easy for a partner to come in and build an output plugin for their own export format.

Reply:
@rgwozdz and I connected and decided that we'll keep this provider confined to performing the Hub search API-related tasks, while steps 1-3 will happen in the DCAT output plugins. HubJS already mitigates the duplication.

We also talked about where the appropriate place to attach the search configuration object is (it's currently at `app.locals`). From the express [doc](http://expressjs.com/en/api.html#res.locals) it looks like the best place would actually be on `res.locals`, but that doesn't work here since the provider plugin won't have access to `res`.

Reply:
Actually, we might be able to get at it through `req.res.locals.` I'll give that a try.

Reply:
As far as I can tell, there isn't any `req.res.locals`, just a `res.locals`.

Reply:
I'm seeing a reference to the response object at `req.res`.

![Screen Shot 2021-07-20 at 11 29 49 AM](https://user-images.githubusercontent.com/315764/126361301-ba3ad70d-32bd-4c99-8d36-fd63ab6add58.png)

Reply:
Ah, ok.  Because, `req` has a reference to `res`.  Oddly, I don't see that in the Express docs. https://expressjs.com/en/4x/api.html#req

Reply:
I don't either, but they do use it in the source: https://github.com/expressjs/express/blob/master/lib/request.js#L469

Reply:
Ah, found it. It gets added in the `init` middleware. https://github.com/expressjs/express/blob/master/lib/middleware/init.js#L31

So, I think it is official.

Reply:
Probably fine, though maybe there aren't any guarantees it will be there in future releases.